File: backend/backend_JAVA/src/main/java/backend/controllers/ProductController.java
package backend.controllers;

import backend.exceptions.ResourceNotFoundException;
import backend.models.ApiResponse;
import backend.models.Product;
import backend.services.ProductService;
import backend.validators.ProductValidator;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.WebDataBinder;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/product")
public class ProductController {
    
    private final ProductService productService;
    private final ProductValidator productValidator;
    
    public ProductController(ProductService productService, ProductValidator productValidator) {
        this.productService = productService;
        this.productValidator = productValidator;
    }
    
    @InitBinder
    protected void initBinder(WebDataBinder binder) {
        binder.addValidators(productValidator);
    }
    
    @GetMapping("/{productId}")
    public ResponseEntity<ApiResponse<Product>> getProductById(@PathVariable String productId) {
        try {
            Product product = productService.getProductById(productId);
            return new ResponseEntity<>(ApiResponse.success(product), HttpStatus.OK);
        } catch (ResourceNotFoundException e) {
            return new ResponseEntity<>(ApiResponse.error("Product not found"), HttpStatus.NOT_FOUND);
        } catch (Exception e) {
            return new ResponseEntity<>(ApiResponse.error("Internal server error"), HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
    
    @GetMapping("/all")
    public ResponseEntity<ApiResponse<List<Product>>> getAllProducts() {
        try {
            List<Product> products = productService.getAllProducts();
            return new ResponseEntity<>(ApiResponse.success(products), HttpStatus.OK);
        } catch (Exception e) {
            return new ResponseEntity<>(ApiResponse.error("Internal server error"), HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
    
    @PostMapping
    public ResponseEntity<ApiResponse<Object>> createProduct(@Valid @RequestBody Product product, BindingResult bindingResult) {
        try {
            if (bindingResult.hasErrors()) {
                Map<String, String> errors = new HashMap<>();
                bindingResult.getFieldErrors().forEach(err -> {
                    errors.put(err.getField(), err.getDefaultMessage());
                });
                return new ResponseEntity<>(ApiResponse.error("Missing required product fields", errors), HttpStatus.BAD_REQUEST);
            }
            
            Product createdProduct = productService.createProduct(product);
            return new ResponseEntity<>(ApiResponse.success(createdProduct), HttpStatus.CREATED);
        } catch (Exception e) {
            return new ResponseEntity<>(ApiResponse.error("Internal server error"), HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
    
    @PutMapping("/{productId}")
    public ResponseEntity<ApiResponse<Object>> updateProduct(@PathVariable String productId, 
                                                            @Valid @RequestBody Product product, 
                                                            BindingResult bindingResult) {
        try {
            if (bindingResult.hasErrors()) {
                Map<String, String> errors = new HashMap<>();
                bindingResult.getFieldErrors().forEach(err -> {
                    errors.put(err.getField(), err.getDefaultMessage());
                });
                return new ResponseEntity<>(ApiResponse.error("Invalid product data", errors), HttpStatus.BAD_REQUEST);
            }
            
            Product updatedProduct = productService.updateProduct(productId, product);
            return new ResponseEntity<>(ApiResponse.success(updatedProduct), HttpStatus.OK);
        } catch (ResourceNotFoundException e) {
            return new ResponseEntity<>(ApiResponse.error("Product not found"), HttpStatus.NOT_FOUND);
        } catch (Exception e) {
            return new ResponseEntity<>(ApiResponse.error("Internal server error"), HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
    
    @DeleteMapping("/{productId}")
    public ResponseEntity<ApiResponse<Object>> deleteProduct(@PathVariable String productId) {
        try {
            boolean result = productService.deleteProduct(productId);
            
            if (result) {
                return new ResponseEntity<>(ApiResponse.success(null, "Product deleted successfully"), HttpStatus.OK);
            } else {
                return new ResponseEntity<>(ApiResponse.error("Product not found"), HttpStatus.NOT_FOUND);
            }
        } catch (Exception e) {
            return new ResponseEntity<>(ApiResponse.error("Internal server error"), HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
} 

File: backend/backend_JS/src/controllers/ProductController.ts
import { Request, Response } from 'express';
import ProductService from '../services/ProductService';

class ProductController {
  /**
   * Get a product by ID
   */
  async getProductById(req: Request, res: Response): Promise<void> {
    try {
      const { productId } = req.params;
      const product = await ProductService.getProductById(productId);
      
      if (!product) {
        res.status(404).json({ success: false, message: 'Product not found' });
        return;
      }
      
      res.status(200).json({ success: true, data: product });
    } catch (error) {
      console.error('Error getting product by ID:', error);
      res.status(500).json({ success: false, message: 'Internal server error' });
    }
  }

  /**
   * Get all products
   */
  async getAllProducts(req: Request, res: Response): Promise<void> {
    try {
      const products = await ProductService.getAllProducts();
      res.status(200).json({ success: true, data: products });
    } catch (error) {
      console.error('Error getting all products:', error);
      res.status(500).json({ success: false, message: 'Internal server error' });
    }
  }

  /**
   * Create a new product
   */
  async createProduct(req: Request, res: Response): Promise<void> {
    try {
      const productData = req.body;
      const product = await ProductService.createProduct(productData);
      res.status(201).json({ success: true, data: product });
    } catch (error) {
      console.error('Error creating product:', error);
      res.status(500).json({ success: false, message: 'Internal server error' });
    }
  }

  /**
   * Update an existing product
   */
  async updateProduct(req: Request, res: Response): Promise<void> {
    try {
      const { productId } = req.params;
      const productData = req.body;
      
      const updatedProduct = await ProductService.updateProduct(productId, productData);
      
      if (!updatedProduct) {
        res.status(404).json({ success: false, message: 'Product not found' });
        return;
      }
      
      res.status(200).json({ success: true, data: updatedProduct });
    } catch (error) {
      console.error('Error updating product:', error);
      res.status(500).json({ success: false, message: 'Internal server error' });
    }
  }

  /**
   * Delete a product
   */
  async deleteProduct(req: Request, res: Response): Promise<void> {
    try {
      const { productId } = req.params;
      
      const result = await ProductService.deleteProduct(productId);
      
      if (!result) {
        res.status(404).json({ success: false, message: 'Product not found' });
        return;
      }
      
      res.status(200).json({ success: true, message: 'Product deleted successfully' });
    } catch (error) {
      console.error('Error deleting product:', error);
      res.status(500).json({ success: false, message: 'Internal server error' });
    }
  }
}

export default new ProductController(); 